### PR TITLE
Updated JandexUtil to new subtyping

### PR DESCRIPTION
**Pull request by FroMage**
So, @geoand @dmlloyd WDYT of this?

I reimplemented it like I explained on zulip, with proper comments.

I think the implementation is easier to understand and optimal in that we only apply the mapping on the way down from a result.

Now, I had to change two tests because I disagreed about the results:

- One was looking up type args of `Foo<T>` from `Bar<S> implements Foo<S>` and used to throw. I'm not sure what the criteria was for throwing, but IMO the type arguments we're looking for are just `[S]`, so I altered the test. CI can tell us if we used it a way that expected the exception.
- One was looking up type args of `Foo<T>` from Bar implements Foo` and used to return `[]`, while IMO this is simple erasure and the type arg is effectively `[Object]`.

Note that while doing this I realised we've zero tests for component types (inner types) or unresolved type variables (not sure how those can happen). But probably it's fine to wait for a bug to be reported.

### CODE REVIEW

**Review by geoand — APPROVED** [review]
Great! I don't have to support this one anymore :wink:

**Comment by FroMage** [comment]
> Great! I don't have to support this one anymore wink

Well, the idea is that anyone can understand this impl, so if that's not the case for you we can keep the current one if you prefer. I mean, I'm biased in that I came up with it, but if it's not simpler in your opinion we can get rid of the PR, no offence :)

**Comment by geoand** [comment]
No offense taken! This implementation is much better than what I came up with, so in any case I'm +1

**Comment by geoand** [comment]
Let's wait for OK from @dmlloyd as well

**Comment by FroMage** [comment]
Sure.

**Comment by dmlloyd** [comment]
I trust your judgement! (Also super busy... unless you absolutely *need* another pair of eyes here...)

**Comment by FroMage** [comment]
OK, let's merge it then.

**Pull request merged**
